gossip/store_decided_state.go

```diff
@@ -17,30 +17,49 @@ type BlockEpochState struct {
 }
 
 func (s *Store) SetHistoryBlockEpochState(epoch idx.Epoch, bs iblockproc.BlockState, es iblockproc.EpochState) {
-	s.rlp.Set(s.table.BlockEpochStateHistory, epoch.Bytes(), BlockEpochState{
+	bes := &BlockEpochState{
 		BlockState: &bs,
 		EpochState: &es,
-	})
+	}
+	// Write to the DB
+	s.rlp.Set(s.table.BlockEpochStateHistory, epoch.Bytes(), bes)
+	// Save to the LRU cache
+	s.cache.BlockEpochStateHistory.Add(epoch, bes, nominalSize)
 }
 
 func (s *Store) GetHistoryBlockEpochState(epoch idx.Epoch) (*iblockproc.BlockState, *iblockproc.EpochState) {
-	// check current BlockEpochState as a cache
-	if v := s.cache.BlockEpochState.Load(); v != nil {
+	// Get HistoryBlockEpochState from LRU cache first.
+	if v, ok := s.cache.BlockEpochStateHistory.Get(epoch); ok {
 		bes := v.(*BlockEpochState)
 		if bes.EpochState.Epoch == epoch {
```

Comment:
I suggest to simplify the code here
```
	if v, ok := s.cache.BlockEpochStateHistory.Get(epoch); ok {
		bes := v.(*BlockEpochState)
		if bes.EpochState.Epoch == epoch {
			return bes.BlockState,bes.EpochState
		}
```

Reply:
By doing this
```			
bs := *bes.BlockState
es := *bes.EpochState
return &bs, &es
```
We want to ensure that returned values have different address, so that caller couldn't accidentally modify the value in cache. Now when I think about it, we should actually call
```			
bs := bes.BlockState.Copy()
es := bes.EpochState.Copy()
return &bs, &es
```
So that .Copy would also make a copy of underlying arrays and pointers